src/xwayland/xwm.rs

```diff
@@ -0,0 +1,476 @@
+            ConfigureWindowAux, ConnectionExt as _, CreateWindowAux, EventMask, Screen, StackMode,
+            Window as X11Window, WindowClass,
+        },
+        Event,
+    },
+    rust_connection::{ConnectionError, DefaultStream, RustConnection},
+    COPY_DEPTH_FROM_PARENT,
+};
+
+x11rb::atom_manager! {
+    Atoms: AtomsCookie {
+        WM_S0,
+        WL_SURFACE_ID,
+        _LATE_SURFACE_ID,
+        _SMITHAY_CLOSE_CONNECTION,
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct X11Surface {
+    window: X11Window,
+    wl_surface: Option<WlSurface>,
+    conn: Weak<RustConnection>,
+    state: Rc<RefCell<SharedSurfaceState>>,
+}
+
+#[derive(Debug)]
+struct SharedSurfaceState {
+    mapped: bool,
+    mapped_onto: X11Window,
+    location: Point<i32, Logical>,
+    size: Size<i32, Logical>,
+}
+
+impl PartialEq for X11Surface {
+    fn eq(&self, other: &Self) -> bool {
+        self.window == other.window && self.wl_surface == other.wl_surface
+    }
+}
+
+impl X11Surface {
+    pub fn set_mapped(&self, mapped: bool) -> Result<(), ConnectionError> {
+        if let Some(conn) = self.conn.upgrade() {
+            let mut state = self.state.borrow_mut();
+            let frame = state.mapped_onto;
+            state.mapped = mapped;
+            if mapped {
+                conn.map_window(frame)?;
+            } else {
+                conn.unmap_window(frame)?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn configure(&mut self, rect: Rectangle<i32, Logical>) -> Result<(), ConnectionError> {
+        if let Some(conn) = self.conn.upgrade() {
+            let aux = ConfigureWindowAux::default()
+                .x(rect.loc.x)
+                .y(rect.loc.y)
+                .width(rect.size.w as u32)
+                .height(rect.size.h as u32);
+            conn.configure_window(self.window, &aux)?;
+            let mut state = self.state.borrow_mut();
+            state.location = rect.loc;
+            state.size = rect.size;
+        }
+        Ok(())
+    }
+
+    pub fn window_id(&self) -> X11Window {
+        self.window
+    }
+
+    pub fn wl_surface(&self) -> Option<&WlSurface> {
+        self.wl_surface.as_ref()
+    }
+
+    pub fn geometry(&self) -> Rectangle<i32, Logical> {
+        let state = self.state.borrow();
+        Rectangle::from_loc_and_size(state.location, state.size)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum X11Request {
+    NewWindow {
+        window: X11Surface,
+        location: Point<i32, Logical>,
+    },
+    DestroyedWindow {
+        window: X11Surface,
+    },
+    Configure {
+        window: X11Surface,
+        x: Option<i32>,
+        y: Option<i32>,
+        width: Option<u32>,
+        height: Option<u32>,
+        border_width: Option<u32>,
+        stacking: Option<(StackMode, X11Window)>,
+    },
+}
+
+/// The runtime state of the XWayland window manager.
+#[derive(Debug)]
+pub struct X11WM {
```

Comment:
On naming are we supposed to use `WM` or `Wm`?

Reply:
Well technically it is an acronym, so `WM`. But I see the appeal for `Wm`, what does you gut say?

Reply:
I'd think `Wm`, as that is in line with how Unix (UnixStream) is used through out the standard library.

Reply:
Sure. I can change that.